source/agora/common/TransactionManager.d

```diff
@@ -0,0 +1,283 @@
+import agora.common.Data;
+import agora.common.crypto.Key;
+import agora.common.Transaction;
+import std.digest.sha;
+import vibe.data.json : serializeToJsonString;
+
+/*******************************************************************************
+
+Class for the transactions verification
+
+*******************************************************************************/
+
+public class TransactionManager
+{
+    /// Store of previously processed transactions
+    private Transaction[Hash] txs;
+
+    /// null value of transaction hash
+    public static Hash nullTxHash;
+
+    /// null value of signature
+    public static Signature nullSignature;
+
+    /*******************************************************************************
+
+        Save a `Transaction`
+
+        Params:
+            hash = `Hash` of `Transaction`
+            tx = Type of `Transaction`
+
+    *******************************************************************************/
+
+    public void saveTransaction (Hash hash, Transaction tx)
+    {
+        this.txs[hash] = tx;
+    }
+
+    /*******************************************************************************
+
+        Return hash of transaction
+
+        Params:
+            tx = Type of `Transaction`
+
+        Return:
+            `Hash` of `Transaction`
+
+    *******************************************************************************/
+
+    public Hash getTxHash (Transaction tx)
+    {
+        Hash hash = sha256Of(serializeToJsonString(tx));
+        return hash;
+    }
+
+    /*******************************************************************************
+
+        New `Input` of `Transaction`
+
+        Params:
+            previous = The hash of a previous transaction containing the `Output` to spend
+            index = Index of the `Output` in the `previous` `Transaction`
+            signature = A signature that should be verified using the `previous[index].address` public key
+
+        Return:
+            The input of the transaction
+
+    *******************************************************************************/
+
+    public Input newTxIn (Hash previous, uint index, Signature signature = TransactionManager.nullSignature)
+    {
+        Input txIn;
+        txIn.previous = previous;
+        txIn.index = index;
+        txIn.signature = signature;
+        return txIn;
+    }
+
+    /*******************************************************************************
+
+        New `Output` of `Transaction`
+
+        Params:
+            address =  The public key that can redeem this output
+            value = The value of this output
+
+        Return:
+            The output of the transaction
+
+    *******************************************************************************/
+
+    public Output newTxOut (PublicKey address, Amount value = 0)
+    {
+        Output txOut;
+        txOut.address = address;
+        txOut.value = value;
+        return txOut;
+    }
```

Comment:
This does not use any of the class' variable. So it could be made `static` to allow it to be called without an instance. Or it could be made a free function.

However, I think it can be removed. Because you can use:
```d
auto txout = TxOut(address, value);
```

the same way you would use:
```d
auto txout = txmanager.newTxOut(address, value);
```

D generate a default constructor for `struct` which is equivalent to:
```d
this(PublicKey _1 = PublicKey.init, Amount _2 = Amount.init)
```

So you can call `TxOut`, `TxOut(address)` and `TxOut(address, amount)`.